File: nim/diamond/diamond.nim
import strutils

proc reverse(input: string): string =
    for s in input:
        result = s & result

proc diamond*(input: char): string =
    const start = 'A'
    let steps = ord(input) - ord(start)
    var lines = newSeq[string](steps * 2 + 1)

    for i in 0..steps:
        var half = " ".repeat(steps - i) & char(int(start) + int(i)) & " ".repeat(i)
        var line = half & half.reverse[1..^1]
        lines[i] = line
        lines[steps * 2 - i] = line

    lines.join("\n") & "\n"

File: lua/diamond/diamond.lua
return function(which)

    local start = ("A"):byte()
    local steps = which:byte() - start
    local lines = {}

    for i = 0, steps, 1 do
        local half = (" "):rep(i) .. string.char(start + i) .. (" "):rep(steps - i)
        local line = half:sub(2):reverse() .. half .. "\n"
        lines[i + 1] = line
        lines[2*steps - i + 1] = line
    end

    return table.concat(lines)

end
